File: android/feature-record/src/main/kotlin/com/bolsillo/feature/record/ui/AmountKeypad.kt
package com.bolsillo.feature.record.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.bolsillo.designsystem.theme.BolsilloTheme
import com.bolsillo.feature.record.R

/**
 * Custom 3×4 integer keypad (1–9, 000, 0, ⌫). The OS soft keyboard is never
 * shown (FR1/§2.1) — keypad is up the moment the screen opens.
 */
@Composable
fun AmountKeypad(
    onDigit: (Char) -> Unit,
    onBackspace: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val rows =
        listOf(
            listOf("1", "2", "3"),
            listOf("4", "5", "6"),
            listOf("7", "8", "9"),
            listOf("000", "0", "⌫"),
        )
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        HintStrip()
        rows.forEach { row ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                row.forEach { key ->
                    KeypadKey(
                        label = key,
                        modifier = Modifier.weight(1f),
                        onClick = {
                            when {
                                key == "⌫" -> onBackspace()
                                key == "000" -> {
                                    onDigit('0')
                                    onDigit('0')
                                    onDigit('0')
                                }
                                else -> onDigit(key[0])
                            }
                        },
                    )
                }
            }
        }
    }
}

@Composable
private fun HintStrip() {
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .clip(BolsilloTheme.shapes.control)
                .background(BolsilloTheme.colors.primaryContainer)
                .padding(horizontal = 14.dp, vertical = 10.dp),
        contentAlignment = Alignment.CenterStart,
    ) {
        Text(
            text = stringResource(R.string.record_keypad_hint),
            color = BolsilloTheme.colors.onPrimaryContainer,
            style = BolsilloTheme.typography.label,
        )
    }
}

@Composable
private fun KeypadKey(
    label: String,
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
) {
    Box(
        modifier =
            modifier
                .height(56.dp)
                .shadow(BolsilloTheme.elevation.key, BolsilloTheme.shapes.control)
                .clip(BolsilloTheme.shapes.control)
                .background(BolsilloTheme.colors.surface)
                .clickable(onClick = onClick)
                .semantics { contentDescription = "keypadKey:$label" },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = label,
            color = BolsilloTheme.colors.textPrimary,
            style = BolsilloTheme.typography.keypadDigit,
        )
    }
}

File: ios/Packages/Bolsillo/Sources/FeatureRecord/Views/AmountKeypad.swift
import SwiftUI
import BolsilloDesignSystem

public struct AmountKeypad: View {
    let onDigit: (String) -> Void
    let onBackspace: () -> Void
    @Environment(\.bolsilloTheme) var theme

    private let rows: [[String]] = [
        ["1", "2", "3"],
        ["4", "5", "6"],
        ["7", "8", "9"],
        ["000", "0", "⌫"],
    ]

    public init(onDigit: @escaping (String) -> Void, onBackspace: @escaping () -> Void) {
        self.onDigit = onDigit
        self.onBackspace = onBackspace
    }

    public var body: some View {
        VStack(spacing: theme.spacing.sm) {
            // Hint strip
            HStack {
                Image(systemName: "sparkles")
                    .font(theme.typography.caption)
                    .foregroundStyle(theme.colors.onPrimaryContainer)
                Text(String(localized: "record.keypad.hint", bundle: .module))
                    .font(theme.typography.caption)
                    .foregroundStyle(theme.colors.onPrimaryContainer)
            }
            .padding(.horizontal, theme.spacing.base)
            .padding(.vertical, theme.spacing.xs)
            .frame(maxWidth: .infinity)
            .background(theme.colors.primaryContainer)
            .cornerRadius(theme.radius.control)

            ForEach(rows, id: \.self) { row in
                HStack(spacing: theme.spacing.sm) {
                    ForEach(row, id: \.self) { key in
                        KeypadButton(
                            key: key,
                            theme: theme,
                            action: {
                                if key == "⌫" { onBackspace() }
                                else { onDigit(key) }
                            }
                        )
                    }
                }
            }
        }
    }
}

private struct KeypadButton: View {
    let key: String
    let theme: BolsilloTheme
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(key)
                .font(theme.typography.keypadDigit)
                .foregroundStyle(theme.colors.textPrimary)
                .frame(maxWidth: .infinity, minHeight: 56)
                .background(theme.colors.surface)
                .cornerRadius(theme.radius.control)
                .shadow(
                    color: theme.elevation.key.color,
                    radius: theme.elevation.key.radius,
                    x: theme.elevation.key.x,
                    y: theme.elevation.key.y
                )
        }
        .buttonStyle(.plain)
    }
}
